include/dr/shp/copy.hpp

```diff
@@ -75,19 +75,24 @@ template <std::forward_iterator InputIt, lib::distributed_iterator OutputIt>
   requires __detail::is_syclmemcopyable<std::iter_value_t<InputIt>,
                                         std::iter_value_t<OutputIt>>
 cl::sycl::event copy_async(InputIt first, InputIt last, OutputIt d_first) {
-  auto segments = lib::ranges::segments(d_first);
-  auto segment = std::begin(segments);
+  auto &&segments = lib::ranges::segments(d_first);
+  auto segment_iter = rng::begin(segments);
 
   std::vector<cl::sycl::event> events;
 
   while (first != last) {
-    const std::size_t n_to_copy =
-        std::min<size_t>((*segment).size(), last - first);
+    auto &&segment = *segment_iter;
+    auto size = std::distance(rng::begin(segment), rng::end(segment));
 
-    events.push_back(
-        shp::copy_async(first, first + n_to_copy, (*segment).begin()));
+    std::size_t n_to_copy = std::min<size_t>(size, std::distance(first, last));
```

Comment:
`size` is a variable used in one place only. If we do not need its name for improve readability of the code I suggest putting `std::distance(rng::begin(segment), rng::end(segment))` inside `min`.
... or making `std::distance(first, last)` also a variable to be consistent (havew something like `min(segment_size, remainder_size`).

But this is just cosmetic. I'm ok with this also .

Reply:
Segments are not necessarily sized, so they might not have `rng::size` or `.size()`.  `std::distance()` will work for any forward range.

At some point we really want our own wrapper that will return `rng::size()` for a sized range and `std::distance()` otherwise.